engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -54,20 +55,53 @@ private void logList(String title, List<?> list) {
     private void generateTaskList() {
```

Comment:
Let's take a step back. The previous version of this method was:

```
private void generateTaskList() {
    taskList.clear();

    Map intranodesStateChanges = Maps.newHashMap();
    for (Object object : intermediateList) {

        if (object instanceof StateChange) {
            intranodesStateChanges.put(object.getClass(), object);
       
        } else {
            for (Object stateChange : intranodesStateChanges.values()) {
                taskList.add(((StateChange) stateChange).generateTask());
            }

            intranodesStateChanges.clear();
            taskList.add(((Node) object).generateTask());

        }
    }
}
```

First let's address the most important problem, the state changes that persist beyond their node's process() execution. For that I would change the method this way:

```
private void generateTaskList() {

    StateChange stateChange;
    StateChange persistentStateChange;
    Map persistentStateChangeS = Maps.newHashMap();  // assuming we can't make it a private field for the time being
    Map intranodesStateChanges = Maps.newHashMap();

    taskList.clear();

    for (Object object : intermediateList) {

        if (object instanceof StateChange) {
            intranodesStateChanges.put(object.getClass(), object);
       
        } else {
            for (Object preCastStateChange : intranodesStateChanges.values()) {
                stateChange = (StateChange) preCastStateChange;

                persistentStateChange = persistentStateChangeS.get(stateChange.getClass);
                if(persistentStateChange == null) {
                    if(!stateChange.isNotTheDefaultInstance()) { // yep, new (convenience) method, just for readability
                        // defensive programming here: the check is probably unnecessary as for every default
                        // instance of a state change subType there should be a non-default one already in the 
                        // persistentStateChangeS map, falling within cases handled below.
                        taskList.add(stateChange.generateTask());
                        persistentStateChangeS.put(stateChange.getClass, stateChange);

                    } else {
                        // do nothing: we do not want a back-to-default state change if there was no known
                        // non-default state change already taking place.
                    }

                }  else {
                   if(stateChange.isTheDefaultInstance()) { // I know, I'm a sucker for almost plain-english code
                        // non redundant default state change, eliminates subType entry in the map, to keep map small
                        taskList.add(stateChange.generateTask());
                        persistentStateChangeS.remove(stateChange.getClass);

                    } else if (stateChange.isNotLike(persistentStateChange)) { // another new method, just for readability
                         // non-redundant state change of the same subType but different value, becomes new map entry
                         taskList.add(stateChange.generateTask());
                         persistentStateChangeS.put(stateChange.getClass, stateChange)

                    } else {
                         // the non-redundant state change being examined is not a default one and is identical
                         // in value to the one stored in the persistentStateChange map: it is redundant after all
                         // and we ignore it
                    }
                }

            }

            intranodesStateChanges.clear();
            taskList.add(((Node) object).generateTask());

        }
    }
}
```

I've written the new version above without the help of the IDE, straight here in GitHub, so there must be errors to fix.
How does it look to you @tdgunes? 

This doesn't address the issue of default state changes after the last node potentially being redundant with the state changes before the first node, but perhaps this is not a bad thing: there might be opengl things happening between frames that are not part of the rendering engine and it's probably best if the rendering engine cleans after itself to leave a default opengl state. So,  unless you have reasons against it,  I'd leave this one out after all.